Applications/V7/cmd/dd.1
.\" UNIX V7 source code: see /COPYRIGHT or www.tuhs.org for details.
.TH DD 1 
.SH NAME
dd \- convert and copy a file
.SH SYNOPSIS
.B dd
[option=value] ...
.SH DESCRIPTION
.I Dd
copies the specified input file
to the specified output with
possible conversions.
The standard input and output are used by default.
The input and output block size may be
specified to take advantage of raw physical I/O.
.PP
.br
.ns
.TP 15
.I option
.I values
.br
.ns
.TP 
if=
input file name; standard input is default
.br
.ns
.TP 
of=
output file name; standard output is default
.br
.ns
.TP 
.RI ibs= n
input block size
.I n
bytes (default 512)
.br
.ns
.TP 
.RI obs= n
output block size (default 512)
.br
.ns
.TP 
.RI bs= n
set both input and output block size,
superseding
.I ibs
and
.I obs;
also, if no conversion is specified,
it is particularly efficient since no copy need be done
.br
.ns
.TP 
.RI cbs= n
conversion buffer size
.br
.ns
.TP 
.RI skip= n
skip
.IR n ""
input records before starting copy
.br
.ns
.TP
.RI files= n
copy
.I n
files from (tape) input
.br
.ns
.TP 
.RI seek= n
seek
.I n
records from beginning of output file before copying
.br
.ns
.TP 
count=\fIn\fR
copy only
.IR n ""
input records
.br
.ns
.TP 
conv=ascii
.ds h \h'\w'conv='u'
convert EBCDIC to ASCII
.br
.ns
.IP \*hebcdic
convert ASCII to EBCDIC
.br
.ns
.IP \*hibm
slightly different map of ASCII to EBCDIC
.br
.ns
.IP \*hlcase
map alphabetics to lower case
.br
.ns
.IP \*hucase
map alphabetics to upper case
.br
.ns
.IP \*hswab
swap every pair of bytes
.br
.ns
.IP \*hnoerror
do not stop processing on an error
.br
.ns
.IP \*hsync
pad every input record to
.I  ibs
.br
.ns
.IP "\*h... , ..."
several comma-separated conversions
.PP
.fi
Where sizes are specified,
a number of bytes is expected.
A number may end with
.B "k, b"
or
.B w
to specify multiplication by
1024, 512, or 2 respectively;
a pair of numbers may be separated by
.B x
to indicate a product.
.PP
.I Cbs
is used only if
.I ascii
or
.I ebcdic
conversion is specified.
In the former case
.I cbs
characters are placed into the conversion buffer, converted to
ASCII, and trailing blanks trimmed and new-line added
before sending the line to the output.
In the latter case ASCII characters are read into the
conversion buffer, converted to EBCDIC, and blanks added
to make up an
output record of size
.IR cbs .
.PP
After completion,
.I dd
reports the number of whole and partial input and output
blocks.
.PP
For example, to read an EBCDIC tape blocked ten 80-byte
EBCDIC card images per record into the ASCII file
.IR x :
.IP ""
dd if=/dev/rmt0 of=x ibs=800 cbs=80 conv=ascii,lcase
.PP
Note the use of raw magtape.
.I Dd
is especially suited to I/O on the raw
physical devices because it allows reading
and writing in arbitrary record sizes.
.PP
To skip over a file before copying from magnetic tape do
.IP""
(dd of=/dev/null; dd of=x) </dev/rmt0
.SH "SEE ALSO"
cp(1), tr(1)
.SH DIAGNOSTICS
f+p records in(out): numbers of full and partial records read(written)
.SH BUGS
The ASCII/EBCDIC conversion tables are
taken
from the 256 character standard in
the CACM Nov, 1968.
The `ibm' conversion, while less blessed as a standard,
corresponds better to certain IBM print train conventions.
There is no universal solution.
.PP
Newlines are inserted only on conversion to ASCII;
padding is done only on conversion to EBCDIC.
These should be separate options.


Applications/util/dd.c
/*
 * Copyright (c) 1993 by David I. Bell
 * Permission is granted to use, distribute, or modify this source,
 * provided that this copyright notice remains intact.
 *
 * The "dd" built-in command.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <fcntl.h>


typedef unsigned char BOOL;

#ifndef FALSE
#define FALSE  0
#define TRUE   1
#endif

#define	PAR_NONE	0
#define	PAR_IF		1
#define	PAR_OF		2
#define	PAR_BS		3
#define	PAR_COUNT	4
#define	PAR_SEEK	5
#define	PAR_SKIP	6


typedef struct {
    char *name;
    int value;
} PARAM;


static PARAM params[] =
{
    {"if",    PAR_IF},
    {"of",    PAR_OF},
    {"bs",    PAR_BS},
    {"count", PAR_COUNT},
    {"seek",  PAR_SEEK},
    {"skip",  PAR_SKIP},
    {NULL,    PAR_NONE}
};


static long getnum(char *);

BOOL intflag;

static char localbuf[8192];

int main(int argc, char *argv[])
{
    char *str;
    char *cp;
    PARAM *par;
    char *infile;
    char *outfile;
    int infd;
    int outfd;
    int incc;
    int outcc;
    int blocksize;
    long count;
    long seekval;
    long skipval;
    long intotal;
    long outtotal;
    long inmax;
    char *buf;
    int ret = 1;

    infile = NULL;
    outfile = NULL;
    seekval = 0;
    skipval = 0;
    blocksize = 512;
    count = 0x7fffffff;
    inmax = 0;

    while (--argc > 0) {
	str = *++argv;
	cp = strchr(str, '=');
	if (cp == NULL) {
	    fprintf(stderr, "Bad dd argument\n");
	    return 1;
	}
	*cp++ = '\0';

	for (par = params; par->name; par++) {
	    if (strcmp(str, par->name) == 0)
		break;
	}

	switch (par->value) {
	case PAR_IF:
	    if (infile) {
		fprintf(stderr, "Multiple input files illegal\n");
		return 1;
	    }
	    infile = cp;
	    break;

	case PAR_OF:
	    if (outfile) {
		fprintf(stderr, "Multiple output files illegal\n");
		return 1;
	    }
	    outfile = cp;
	    break;

	case PAR_BS:
	    blocksize = getnum(cp);
	    if (blocksize <= 0) {
		fprintf(stderr, "Bad block size value\n");
		return 1;
	    }
	    break;

	case PAR_COUNT:
	    count = getnum(cp);
	    if (count < 0) {
		fprintf(stderr, "Bad count value\n");
		return 1;
	    }
	    break;

	case PAR_SEEK:
	    seekval = getnum(cp);
	    if (seekval < 0) {
		fprintf(stderr, "Bad seek value\n");
		return 1;
	    }
	    break;

	case PAR_SKIP:
	    skipval = getnum(cp);
	    if (skipval < 0) {
		fprintf(stderr, "Bad skip value\n");
		return 1;
	    }
	    break;

	default:
	    fprintf(stderr, "Unknown dd parameter\n");
	    return 1;
	}
    }

    if (infile == NULL) {
	fprintf(stderr, "No input file specified\n");
	return 1;
    }
    if (outfile == NULL) {
	fprintf(stderr, "No output file specified\n");
	return 1;
    }
    buf = localbuf;
    if (blocksize > sizeof(localbuf)) {
	buf = malloc(blocksize);
	if (buf == NULL) {
	    fprintf(stderr, "Cannot allocate buffer\n");
	    return 1;
	}
    }
    intotal = 0;
    outtotal = 0;

    infd = open(infile, 0);
    if (infd < 0) {
	perror(infile);
	if (buf != localbuf)
	    free(buf);
	return 1;
    }
    outfd = creat(outfile, 0666);
    if (outfd < 0) {
	perror(outfile);
	close(infd);
	if (buf != localbuf)
	    free(buf);
	return 1;
    }
    if (skipval) {
	if (lseek(infd, skipval * blocksize, 0) < 0) {
	    while (skipval-- > 0) {
		incc = read(infd, buf, blocksize);
		if (incc < 0) {
		    perror(infile);
		    goto cleanup;
		}
		if (incc == 0) {
		    fprintf(stderr, "End of file while skipping\n");
		    goto cleanup;
		}
	    }
	}
    }
    if (seekval) {
	if (lseek(outfd, seekval * blocksize, 0) < 0) {
	    perror(outfile);
	    goto cleanup;
	}
    }
    if(count != 0x7fffffff)
        inmax = count * blocksize;
    while ((incc = read(infd, buf, blocksize)) > 0) {
	intotal += incc;
	cp = buf;

	if (intflag) {
	    fprintf(stderr, "Interrupted\n");
	    goto cleanup;
	}
	while (incc > 0) {
	    outcc = write(outfd, cp, incc);
	    if (outcc < 0) {
		perror(outfile);
		goto cleanup;
	    }
	    outtotal += outcc;
	    cp += outcc;
	    incc -= outcc;
	}
        if(inmax && intotal >= inmax)
            break;
    }

    if (incc < 0) {
	perror(infile);
	goto cleanup;
    }

    ret = 0;

  cleanup:
    close(infd);

    if (close(outfd) < 0) {
	perror(outfile);
	ret = 1;
    }

    if (buf != localbuf)
	free(buf);

    printf("%d+%d records in\n", intotal / blocksize,
	   (intotal % blocksize) != 0);

    printf("%d+%d records out\n", outtotal / blocksize,
	   (outtotal % blocksize) != 0);

    return ret;
}


/*
 * Read a number with a possible multiplier.
 * Returns -1 if the number format is illegal.
 */
static long getnum(char *cp)
{
    long value;

    if (!isdigit(*cp))
	return -1;

    value = 0;
    while (isdigit(*cp))
	value = value * 10 + *cp++ - '0';

    switch (*cp++) {
    case 'k':
	value *= 1024;
	break;

    case 'b':
	value *= 512;
	break;

    case 'w':
	value *= 2;
	break;

    case '\0':
	return value;

    default:
	return -1;
    }

    if (*cp)
	return -1;

    return value;
}


/* END CODE */
